x-pack/platform/plugins/shared/cases/server/services/incremental_id/index.ts

```diff
@@ -0,0 +1,391 @@
+import pRetry from 'p-retry';
+import { CASE_SAVED_OBJECT, CASE_ID_INCREMENTER_SAVED_OBJECT } from '../../../common/constants';
+import type { CasePersistedAttributes } from '../../common/types/case';
+import type {
+  CaseIdIncrementerPersistedAttributes,
+  CaseIdIncrementerSavedObject,
+} from '../../common/types/id_incrementer';
+
+type GetCasesParameters = Pick<
+  SavedObjectsFindOptions,
+  'sortField' | 'sortOrder' | 'perPage' | 'page' | 'filter' | 'namespaces'
+>;
+
+export class CasesIncrementalIdService {
+  static incrementalIdExistsFilter = 'cases.attributes.incremental_id: *';
+  static incrementalIdMissingFilter = 'not cases.attributes.incremental_id: *';
+  private isStopped = false;
+
+  constructor(
+    private internalSavedObjectsClient: SavedObjectsClientContract,
+    private logger: Logger
+  ) {
+    this.logger = logger.get('incremental_id_service');
+    this.logger.debug('Cases incremental ID service initialized');
+  }
+
+  public stopService() {
+    this.isStopped = true;
+  }
+  public startService() {
+    this.isStopped = false;
+  }
+
+  public async getCasesWithoutIncrementalId(parameters: Omit<GetCasesParameters, 'filter'> = {}) {
+    return this.getCases({
+      ...parameters,
+      filter: CasesIncrementalIdService.incrementalIdMissingFilter,
+    });
+  }
+
+  public async getCases({
+    filter,
+    perPage = 10000,
+    page = 1,
+    sortOrder = 'asc',
+    sortField = 'created_at',
+    namespaces = ['*'],
+  }: GetCasesParameters) {
+    try {
+      const savedCases = await this.internalSavedObjectsClient.find<CasePersistedAttributes>({
+        type: CASE_SAVED_OBJECT,
+        sortField,
+        sortOrder,
+        perPage,
+        page,
+        filter,
+        namespaces,
+      });
+      return savedCases;
+    } catch (error) {
+      this.logger.error(error);
+      throw error;
+    }
+  }
+
+  /**
+   * Get the latest applied ID for a given space.
+   * Uses the actually applied numerical ids on cases in the space.
+   */
+  public async getLastAppliedIdForSpace(namespace: string) {
+    try {
+      const casesResponse = await this.getCases({
+        filter: CasesIncrementalIdService.incrementalIdExistsFilter,
+        namespaces: [namespace],
+        sortField: 'incremental_id',
+        sortOrder: 'desc',
+        perPage: 1, // We only need the most recent incremental id value
+        page: 1,
+      });
+
+      if (casesResponse.total === 0) {
+        this.logger.debug(`No cases found with incremental id in ${namespace}`);
+        return 0;
+      }
+
+      const mostRecentIncrementalId = casesResponse.saved_objects[0].attributes.incremental_id;
+      this.logger.debug(
+        `getLastAppliedIdForSpace (from cases): Most recent incremental id in ${namespace}: ${mostRecentIncrementalId}`
+      );
+
+      if (mostRecentIncrementalId === undefined || mostRecentIncrementalId === null) {
+        return 0;
+      }
+
+      // TODO: should we really throw here?
```

Comment:
I don't think we should throw here, but this could be a good telemetry point for how often this takes place, but still thinking through the best structure for the telemetry in general for this service

Reply:
This comment was outdated actually ffdb664e0e4